File: storage/data_original/at_coder/abc054/C/4907923.py
#??????????????????????????????????(??????????????????)
#???????????????
#????????????()????????????????DFS?BFS?????????O(n^2)???O(n!)??????
from itertools import permutations

N,M = map(int,input().split())
E = []
for i in range(M):
    a,b = (int(_) for _ in input().split())
    E.append((a,b))
V = (i+1 for i in range(N))
one_strike_cnt = 0

for vertexs in permutations(V):
    #print(vertexs)
    if vertexs[0] != 1:
        continue
    #print(vertexs)
    one_strike = True#?????????True
    for vertex_index in range(len(vertexs)-1):
        flag = False
        for i in range(M):
            a,b = E[i]
            if (a == vertexs[vertex_index] and b == vertexs[vertex_index+1]) or (a == vertexs[vertex_index+1] and b == vertexs[vertex_index]):
                flag = True
                break
        if not flag:
            one_strike = False
            break
    if one_strike:
        one_strike_cnt += 1

print(one_strike_cnt)

File: data/data_transpiled/at_coder/abc054/C/4907923.java
@ VisibleForTesting static Iterable < Integer > permutations ( ) {
  int N = Integer . parseInt ( input . nextLine ( ) ) ;
  int M = Integer . parseInt ( input . nextLine ( ) ) ;
  List < Integer > E = new ArrayList < > ( ) ;
  for ( int i = 0 ;
  i < M ;
  i ++ ) {
    int a = ( Integer ) input . nextLine ( ) ;
    int b = ( Integer ) input . nextLine ( ) ;
    E . add ( new Integer ( a ) ) ;
  }
  int [ ] V = new int [ N ] ;
  for ( int i = 0 ;
  i < N ;
  i ++ ) {
    V [ i ] = i + 1 ;
  }
  int oneStrikeCnt = 0 ;
  for ( int [ ] vertexs : permutations ( V ) ) {
    if ( vertexs [ 0 ] != 1 ) {
      continue ;
    }
    boolean oneStrike = true ;
    for ( int vertexIndex = 0 ;
    vertexIndex < vertexs . length - 1 ;
    vertexIndex ++ ) {
      boolean flag = false ;
      for ( int i = 0 ;
      i < M ;
      i ++ ) {
        int a = E . get ( i ) ;
        int b = E . get ( i ) ;
        if ( ( a == vertexs [ vertexIndex ] && b == vertexs [ vertexIndex + 1 ] ) || ( a == vertexs [ vertexIndex + 1 ] && b == vertexs [ vertexIndex ] ) ) {
          flag = true ;
          break ;
        }
      }
      if ( ! flag ) {
        oneStrike = false ;
        break ;
      }
    }
    if ( oneStrike ) {
      oneStrikeCnt ++ ;
    }
  }
  return E ;
}
